docs/proposals/004-scalability-benchmarking.md

```diff
@@ -0,0 +1,102 @@
+2. Determine the baseline for when tweaking is required on the default configuration (resource allocations and replicas).
+    1. One cluster, Applications in In-cluster, default resource allocations.
+3. Quantify how tweaking of existing parameters (replicas, sharding, parallelism, etc) impacts the performance of Argo CD.
+4. Provide a set of metrics and thresholds that will provide a basis for automatically scaling the Argo CD components and alerting for when performance is being impacted by limitations.
+
+### Non-Goals
+* This proposal does not intend to cover implementation of any auto-scaling based on the metrics and thresholds determined by the scalability benchmarking. A separate proposal will be used for any auto-scaling enhancements.
+* This proposal does not intend to add testing that determines how a change impacts the scalability of Argo CD based on the benchmarks.
+* We do not intend to analyze cost implications of running different topologies and purely focus on scalability requirements from a technology perspective.
+
+## Proposal
+1.  Create new `argo-cd-benchmarking` repo under `argoproj-labs` and add authors of this proposal as maintainers.
+2. Create a set of key scalability factors to use as testing parameters. For example:
+    1. Number of Applications.
+    2. Number of resources managed by an Application.
+    3. Number of resources in a cluster.
+    4. The size of the resources in a cluster and managed by an Application.
+    5. Churn rate for resources in the cluster (how often resources change).
+    6. Number of clusters.
+    7. Number of repositories being monitored.
+    8. Size of the repositories.
+3. Determine the metrics that reflect limitations in scalability factors.
+    1. Application sync time for x number of apps
+    2.. Emptying the queues (app_reconciliation_queue, app_operation_processing_queue)
+4. Create automated testing procedures for Argo CD that take the key scalability factors as testing parameters.
+5. Test the default installation of Argo CD to determine the limit based on the key scalability factors.
+6. Create test scenarios that reflect the common topologies (Argo CD 1-1 with clusters, Argo CD 1-many with clusters).
+7. Determine the thresholds for the metrics identified earlier to capture when performance is being impacted.
+    1. Contribute back grafana thresholds and alerts for prometheus
+
+### Use cases
+Each use case will cover a specific topology with N permutations based on the key scalability factors. They will be measured using the metrics given in the proposal.
+
+We intend to focus on two key topologies: one Argo CD per cluster and one Argo CD for all remote clusters. All other variations are an extension of these.
+
+#### Topology 1: One Argo CD per cluster.
+The exact key scalability factors used in each permutations will be determined once we get to the testing.
+
+#### Topology 2: One Argo CD for all remote clusters.
```

Comment:
Topology suggestion:
1. Run ArgoCD in a namespace, and have it manage a couple of other namespaces - but not anything cluster-scoped.

Reply:
Good point! This would be great for determining the impact of monitoring cluster-scoped resources (maybe related to the effect of resource churn in the cluster)